### PR TITLE
Preserve the base link transform.

### PR DESCRIPTION
**Pull request by ziyan**
1. Fixes a bug where OpenRAVE ignores base link transform when loading JSON
2. Add ability to preserve base link transform for `ExtractInfo` and `UpdateFromKinBodyInfo`

### CODE REVIEW

File: include/openrave/kinbody.h [hunk 1]
```diff
@@ -3481,6 +3481,7 @@ class OPENRAVE_API KinBody : public InterfaceBase
     std::string _id; ///< unique id of the KinBody
     std::string _referenceUri; ///< reference uri saved from InitFromInfo
     boost::shared_ptr<rapidjson::Document> _prAssociatedFileEntries; ///< files tag maintaining entries of data files associated with this object
+    Transform _baseLinkRelativeTransform; ///< the relative transform between body frame and base link, this transform is built into GetTransform() but we need to store it separately for ExtractInfo() to get back the original base link transform
```
Comment (rdiankov) [inline include/openrave/kinbody.h:3484]: Comment not clear. Transform of the base link the body coordinate system?

File: src/libopenrave-core/environment-core.h [hunk 2]
```diff
@@ -3002,14 +3002,14 @@ class Environment : public EnvironmentBase
                 }
 
                 if( !bChanged ) {
-                    dReal dist = TransformDistanceFast(pInitBody->GetTransform(), pKinBodyInfo->_transform);
+                    dReal dist = TransformDistanceFast(pInitBody->GetTransform(), pKinBodyInfo->_transform * pInitBody->_baseLinkRelativeTransform);
```
Comment (rdiankov) [inline src/libopenrave-core/environment-core.h:3005]: err... doesn't make sense, shouldn't `pKinBodyInfo->_transform` be the body frame? So not sure why you have to change here

File: src/libopenrave-core/environment-core.h [hunk 3]
```diff
@@ -3002,14 +3002,14 @@ class Environment : public EnvironmentBase
...
                     if( dist > 1e-7 ) {
                         bChanged = true;
                     }
                 }
 
                 if( bChanged ) {
-                    pInitBody->SetDOFValues(vDOFValues, pKinBodyInfo->_transform, KinBody::CLA_Nothing);
+                    pInitBody->SetDOFValues(vDOFValues, pKinBodyInfo->_transform * pInitBody->_baseLinkRelativeTransform, KinBody::CLA_Nothing);
```
Comment (rdiankov) [inline src/libopenrave-core/environment-core.h:3012]: err... doesn't make sense, shouldn't `pKinBodyInfo->_transform` be the body frame? So not sure why you have to change here

File: src/libopenrave/kinbody.cpp [hunk 4]
```diff
@@ -5984,11 +5987,12 @@ void KinBody::ExtractInfo(KinBodyInfo& info)
     info._vGrabbedInfos.resize(0);
     GetGrabbedInfo(info._vGrabbedInfos);
 
-    info._transform = GetTransform();
+    // GetTransform() = info._transform * _baseLinkRelativeTransform
+    info._transform = GetTransform() * _baseLinkRelativeTransform.inverse();
```
Comment (rdiankov) [inline src/libopenrave/kinbody.cpp:5991]: err... doesn't make sense, shouldn't `info._transform` should be the body frame that is returned by GetTransform()

File: src/libopenrave/kinbody.cpp [hunk 5]
```diff
@@ -5984,11 +5987,12 @@ void KinBody::ExtractInfo(KinBodyInfo& info)
...
 
     // in order for link transform comparision to make sense
     KinBody::KinBodyStateSaver stateSaver(shared_kinbody(), Save_LinkTransformation);
-    SetTransform(Transform());
+    SetTransform(_baseLinkRelativeTransform); // so that base link is at exactly _baseLinkRelativeTransform
```
Comment (rdiankov) [inline src/libopenrave/kinbody.cpp:5995]: why not keep it as `Transform()`?

File: src/libopenrave/kinbody.cpp [hunk 6]
```diff
@@ -6154,7 +6163,7 @@ UpdateFromInfoResult KinBody::UpdateFromKinBodyInfo(const KinBodyInfo& info)
             // if any link has its transform field set, we need to set zero configuration before comparison
             if( (*itLinkInfo)->IsModifiedField(KinBody::LinkInfo::LIF_Transform) ) {
                 stateSaver.reset(new KinBody::KinBodyStateSaver(shared_kinbody(), Save_LinkTransformation));
-                SetTransform(Transform());
+                SetTransform(_baseLinkRelativeTransform); // so that base link is at exactly _baseLinkRelativeTransform
```
Comment (rdiankov) [inline src/libopenrave/kinbody.cpp:6166]: why not keep it as `Transform()`?

File: src/libopenrave/kinbody.cpp [hunk 7]
```diff
@@ -6193,10 +6202,13 @@ UpdateFromInfoResult KinBody::UpdateFromKinBodyInfo(const KinBodyInfo& info)
     }
 
     // transform
-    if( info.IsModifiedField(KinBodyInfo::KBIF_Transform) && GetTransform().CompareTransform(info._transform, g_fEpsilon) ) {
-        SetTransform(info._transform);
-        updateFromInfoResult = UFIR_Success;
-        RAVELOG_VERBOSE_FORMAT("body %s updated due to transform change", _id);
+    if( info.IsModifiedField(KinBodyInfo::KBIF_Transform) || (info._vLinkInfos.size() > 0 && info._vLinkInfos[0]->IsModifiedField(KinBody::LinkInfo::LIF_Transform)) ) {
+        Transform bodyTransform = info._transform * _baseLinkRelativeTransform;
```
Comment (rdiankov) [inline src/libopenrave/kinbody.cpp:6206]: Umm... same thing `GetTransform()` == `info._transform`

**Comment by rdiankov** [comment]
Just made commits to fix all the issues, but I think there's a lot more places that have to be changed in the user code since we sort of assumed GetLinks()[0]->GetTransform() == GetTransform()

**Comment by ziyan** [comment]
@rdiankov The main difference here that I can tell is: I was trying to preserve `GetLinks()[0]->GetTransform() == GetTransform()`, whereas your new changes are saying `GetTransform()` is the body frame in the world transform, and `GetLinks()[0]->GetTransform()` is the base link frame in the world transform.

```json
{"bodies": [{
  "id": "test",
  "name": "test",
  "transform": [1, 0, 0, 0, 1, 0, 0],
  "links": [
    {
      "id": "link0", "name": "L0",
      "transform": [1, 0, 0, 0, 0, 1, 0]
    },
    {
      "id": "link1", "name": "L1",
      "transform": [1, 0, 0, 0, 0, 0, 1]
    }
  ],
  "joints": [
    {
      "id": "joint0", "name": "J0",
      "parentLinkName": "L0", "childLinkName": "L1"
    }
  ]
}]}
```

- `GetLinks()[0]->GetTransform()` should return `[1, 0, 0, 0, 1, 0, 0]` * `[1, 0, 0, 0, 0, 1, 0]`
- `GetLinks()[1]->GetTransform()` should return `[1, 0, 0, 0, 1, 0, 0]` * `[1, 0, 0, 0, 0, 0, 1]`

With your change, I should see:

- `GetTransform()` return `[1, 0, 0, 0, 1, 0, 0]`

(I am still compiling openrave)

File: src/libopenrave-core/environment-core.h (hunk 2 above)
Comment (ziyan) [reply src/libopenrave-core/environment-core.h]: Originally, I was trying to preserve `GetLinks()[0]->GetTransform() == GetTransform()`. So I needed to make `SetTransform` account for `body in world` + `base link in body`.

File: src/libopenrave-core/environment-core.h (hunk 3 above)
Comment (ziyan) [reply src/libopenrave-core/environment-core.h]: Originally, I was trying to preserve `GetLinks()[0]->GetTransform() == GetTransform()`. So I needed to make `SetTransform` account for `body in world` + `base link in body`.

File: src/libopenrave/kinbody.cpp (hunk 6 above)
Comment (ziyan) [reply src/libopenrave/kinbody.cpp]: Originally, I was trying to preserve `GetLinks()[0]->GetTransform() == GetTransform()`. So I needed to put the body temporarily at `base link in body frame` transform, such that the `GetLinks()[0]->GetTransform()` would then be equal to the original `base link in body frame` transform when things are not changed.

File: src/libopenrave/kinbody.cpp (hunk 5 above)
Comment (ziyan) [reply src/libopenrave/kinbody.cpp]: Originally, I was trying to preserve `GetLinks()[0]->GetTransform() == GetTransform()`. So I needed to put the body temporarily at `base link in body frame` transform, such that the `GetLinks()[0]->GetTransform()` would then be equal to the original `base link in body frame` transform when things are not changed.

**Comment by rdiankov** [comment]
any issues with this branch? or has not got tested yet?

**Pull request merged**